matrix/SparseVector.hpp

```diff
@@ -0,0 +1,57 @@
+#include "Vector.hpp"
+#include <cassert>
+
+template<int N> struct force_constexpr_eval {
+    static const int value = N;
+};
+
+// Vector that only store nonzero elements,
+// which indices are specified as parameter pack
+template<typename Type, int... Idxs>
+class SparseVector {
+private:
+    static constexpr size_t N = sizeof...(Idxs);
+    static constexpr int _indices[N] {Idxs...};
+    matrix::Vector<Type, N> _data {};
+
+    static constexpr int findCompressedIndex(int index, int range = N) {
+        const int last_elem = range - 1;
+        return (last_elem < 0 || _indices[last_elem] == index) ? last_elem : findCompressedIndex(index, range - 1);
+    }
+
+public:
+    static constexpr int non_zeros() {
+        return N;
+    }
+    constexpr int index(int i) const {
+        return SparseVector::_indices[i];
+    }
+
+    SparseVector() = default;
+
+    SparseVector(const matrix::Vector<Type, N>& data) : _data(data) {}
```

Comment:
I worry this could be confusing since it remaps the indices of the dense vector. But the other way would also be confusing. Perhaps it would be better to have named initializers to explain that it is doing this, instead of using an un-named constructor?

Reply:
Is it remapping indices? data is already of the size of the sparse vector.

Reply:
The existing behavior makes sense as long as this is just used for direct indexing, but looking forwards we will also want to do things like denseVector.dot(sparseVector), denseMatrix * sparseVector, denseVector += sparseVector, etc. These operations effectively treat the unpopulated indices as 0, but otherwise behave exactly like a dense vector with the same values in it.

Once we have these operators, I would expect `denseMatrix*denseVector` to give the same results as `sparseVector=denseVector; denseMatrix*sparseVector` assuming the denseVector only had non-zeros in the places the sparseVector could hold them.

But I also get that this might not be clear, and different people would have different opinions, so named initializers would make it obvious to anyone what is happening during an initialization. Something like `sparseVector.fromDense(denseVector)` and `sparseVector.initializeInternalRepresentation(denseVector)`

Some of this is coming from working extensively with Eigen's SparseMatrix / SparseVector structures. These are a bit different in that they detect and track sparsity dynamically, which won't work for us. But eventually we will want our SparseVector to be just as useful as Vector, and when that happens it should behave exactly like Vector except for keeping 0 implicitly for the unlisted indices.

Reply:
Thanks for the explanation. This sounds great. Feel free to add it on top of this PR.

Reply:
Will do